dadi/lib/providers/restapi.js

```diff
@@ -0,0 +1,240 @@
+            'Rest provider: cache data incomplete, making HTTP request: ' +
+              err +
+              '(' +
+              cacheOptions.cacheKey +
+              ')'
+          )
+        }
+      }
+
+      var self = this
+
+      this.purest
+        .select(endpoint)
+        .where(query)
+        .request(function (err, res, body) {
+          log.info(
+            { module: 'restapi' },
+            'GOT datasource "' +
+              self.datasource.schema.datasource.key +
+              '": ' +
+              decodeURIComponent(res ? res.request.href : '') +
+              ' (HTTP 200, ' +
+              help.formatBytes(Buffer.byteLength(body)) +
+              ')'
+          )
+
+          self.processOutput(res, err, body, noCache, done)
+        })
+    })
+  } catch (ex) {
+    done(ex, null)
+  }
+}
+
+/**
+ * processOutput
+ *
+ * @param  {response} res
+ * @param  {string} data
+ * @param  {fn} done
+ * @return {void}
+ */
+RestApi.prototype.processOutput = function processOutput (
+  res,
+  error,
+  data,
+  noCache,
+  done
+) {
+  if (res.statusCode === 200) {
+    data = this.processFields(data)
+
+    var cacheOptions = {
+      name: this.datasource.name,
+      caching: this.schema.datasource.caching,
+      cacheKey: this.cacheKey
+    }
+
+    if (!noCache) {
+      this.dataCache.cacheResponse(cacheOptions, JSON.stringify(data), () => {
+        //
+      })
+    }
+  }
+
+  if (res.statusCode === 404) {
+    data = {
+      results: [],
+      errors: [
+        {
+          code: 'WEB-0004',
+          title: 'Datasource Not Found',
+          details:
+            'Datasource "' +
+            this.datasource.name +
+            '" failed. ' +
+            res.statusMessage +
+            ' (' +
+            res.statusCode +
+            ')' +
+            ': ' +
+            res.request.href
+        }
+      ]
+    }
+  }
+
+  // if the error is anything other than Success or Bad Request, error
+  if (res.statusCode === 500) {
+    data = {
+      results: [],
+      errors: [
+        {
+          code: 'WEB-0004',
+          title: 'Datasource Internal Server Error',
+          details:
+            'Datasource "' +
+            this.datasource.name +
+            '" failed. ' +
+            res.statusMessage +
+            ' (' +
+            res.statusCode +
+            ')' +
+            ': ' +
+            res.request.href
+        }
+      ]
+    }
+  }
+
+  if (Buffer.isBuffer(data)) {
+    data = data.toString()
+  }
+
+  if (typeof data === 'string') {
+    data = JSON.parse(data)
+  }
+
+  return done(null, data)
+}
+
+/**
+ * processFields - remove any unwanted fields from the dataset
+ *
+ * @param  {obj} data before it's been processed
+ * @return {obj} data after it's been processed
+ */
+RestApi.prototype.processFields = function processFields (data) {
+  const fields = this.schema.datasource.fields
+
+  if (fields && Object.keys(fields).length) {
+    const keys = Object.keys(fields)
```

Comment:
Could you move this to line 214 and use the value in the `if` expression, to avoid running `Object.keys` twice?